File: dispatcher/src/main/java/shine/Sender.java
package shine;

import java.io.IOException;
import java.net.*;
import java.nio.*;
import java.nio.channels.*;
import java.util.concurrent.atomic.*;

public class Sender {
    InetSocketAddress dest_addr;
    int clientID;
    SocketChannel clientSocket;
    ByteBuffer buffer_A;
    ByteBuffer buffer_B;

    long time_threshold;   // by ms
    long size_threshold;   // by byte

    AtomicBoolean run;
    Thread senderRunner;

    Sender(InetSocketAddress dest_addr, int id) {
        this.time_threshold = 100;
        this.size_threshold = 512 * 1024;
        this.buffer_A = ByteBuffer.allocateDirect(1024 * 1024);
        this.buffer_B = ByteBuffer.allocateDirect(1024 * 1024);
        this.senderRunner = new Thread(new SenderRunner(this));
        run = new AtomicBoolean(true);
        this.dest_addr = dest_addr;
        this.clientSocket = null;
        this.clientID = id;
    }

    public void start_runner() {
        run.set(true);
        senderRunner.start();
    }

    synchronized public void stop_runner() {
        run.set(false);
        try {
            notify();
            senderRunner.join();
        } catch (InterruptedException e) {
            System.out.println(e);
        }

    }

    public void connect() throws IOException {
        clientSocket = SocketChannel.open(dest_addr);
        clientSocket.configureBlocking(true);
        // send greeting message
        ByteBuffer buffer = ByteBuffer.allocateDirect(12);
        buffer.putInt(0xAAAABBBB);
        buffer.putInt(4);
        buffer.putInt(clientID);
        buffer.flip();
        clientSocket.write(buffer);
        buffer.clear();
        clientSocket.read(buffer);
        buffer.flip();
        int response_code = buffer.getInt();
        if (response_code != 200) {
            clientSocket.close();
            clientSocket = null;
            throw new IOException("Got wrong response code.");
        } else {
            System.out.println("Successfully connectec to Combiner server: " + dest_addr);
        }
    }

    public void close() {
        try {
            if (clientSocket != null) {
                clientSocket.close();
            }
        } catch (IOException e) {
            System.out.println(e);
        } finally {
            clientSocket = null;
        }
    }

    synchronized public void send(long identifier, byte[] data, int size) {
        if (buffer_A.remaining() < 16 + data.length) {
            System.out.println("WARNING: buffer is full.");
            return;
        }
        buffer_A.putInt(0xABCDEEFF);
        buffer_A.putInt(8 + size);
        buffer_A.putLong(identifier);
        buffer_A.put(data, 0, size);
        if (buffer_A.position() > size_threshold) {
            notify();
        }
    }

    synchronized public boolean swap() {
        try {
            wait(time_threshold);
        } catch(Exception e) {
            System.out.println(e);
            return false;
        }
        if (buffer_A.position() > 0) {
            ByteBuffer tmp_buffer = buffer_B;
            buffer_B = buffer_A;
            buffer_A = tmp_buffer;
            return true;
        } else {
            return false;
        }
    }
}

class SenderRunner implements Runnable {
    Sender sender;

    SenderRunner(Sender sender) {
        this.sender = sender;
    }

    public void run() {
        while (sender.run.get()) {
            if (sender.swap()) {
                send_buffer(sender.buffer_B);
            }
        }
    }

    void send_buffer(ByteBuffer buffer) {
        if (sender.clientSocket == null || !sender.clientSocket.isConnected()) {
            try {
                sender.connect();
            } catch (IOException e) {
                System.out.println(e);
                sender.close();
                return;
            }
        } else {
            try {
                buffer.flip();
                sender.clientSocket.write(buffer);
                buffer.clear();
            } catch (IOException e) {
                System.out.println(e);
                sender.close();
                return;
            }
        }
    }
}

File: diffraflow/dispatcher/src/main/cpp/Sender.cc
#include "Sender.hh"
#include "PrimitiveSerializer.hh"

#include <sys/socket.h>
#include <arpa/inet.h>
#include <unistd.h>
#include <netdb.h>
#include <string.h>
#include <algorithm>
#include <chrono>

#include <boost/log/trivial.hpp>

diffraflow::Sender::Sender(string hostname, int port, int id) {
    dest_host_ = hostname;
    dest_port_ = port;
    sender_id_ = id;
    buffer_size_ = 1024 * 1024 * 4; // 4 MiB
    size_threshold_ = 1024 * 1024;  // 1 MiB
    time_threshold_ = 100; // 0.1 second
    buffer_A_ = new char[buffer_size_];
    buffer_A_limit_ = 0;
    buffer_B_ = new char[buffer_size_];
    buffer_B_limit_ = 0;
    client_sock_fd_ = -1;
}

diffraflow::Sender::~Sender() {
    delete [] buffer_A_;
    delete [] buffer_B_;
}

bool diffraflow::Sender::connect_to_combiner() {
    addrinfo hints, *infoptr;
    memset(&hints, 0, sizeof(hints));
    hints.ai_family = AF_INET;
    int result = getaddrinfo(dest_host_.c_str(), NULL, &hints, &infoptr);
    if (result) {
        BOOST_LOG_TRIVIAL(error) << "getaddrinfo: " << gai_strerror(result);
        return false;
    }
    client_sock_fd_ = socket(AF_INET, SOCK_STREAM, 0);
    if (client_sock_fd_ < 0) {
        BOOST_LOG_TRIVIAL(error) << "Socket creationg error";
        return false;
    }
    if (connect(client_sock_fd_, infoptr->ai_addr, infoptr->ai_addrlen)) {
        close(client_sock_fd_);
        client_sock_fd_ = -1;
        BOOST_LOG_TRIVIAL(error) << "Connection to " << dest_host_ << " failed.";
        return false;
    }
    freeaddrinfo(infoptr);
    // send greeting message for varification
    char buffer[12];
    gPS.serializeValue<uint32_t>(0xAAAABBBB, buffer, 12);
    gPS.serializeValue<uint32_t>(4, buffer + 4, 8);
    gPS.serializeValue<int32_t>(sender_id_, buffer + 8, 4);
    write(client_sock_fd_, buffer, 12);
    read(client_sock_fd_, buffer, 12);
    int response_code = 0;
    gPS.deserializeValue<int32_t>(&response_code, buffer, 12);
    if (response_code != 200) {
        close(client_sock_fd_);
        client_sock_fd_ = -1;
        BOOST_LOG_TRIVIAL(error) << "Got wrong response code, close the connection.";
        return false;
    } else {
        BOOST_LOG_TRIVIAL(info) << "Successfully connectec to Combiner server " << dest_host_ << ":" << dest_port_;
        return true;
    }
}

void diffraflow::Sender::close_connection() {
    if (client_sock_fd_ >= 0) {
        close(client_sock_fd_);
        client_sock_fd_ = -1;
    }
}

void diffraflow::Sender::push(long key, char* data, size_t len) {
    unique_lock<mutex> lk(mtx_);
    cv_push_.wait(lk, [&]() {return len <= buffer_size_ - buffer_A_limit_;});
    std::copy(data, data + len, buffer_A_);
    buffer_A_limit_ += len;
    if (buffer_A_limit_ > size_threshold_) {
        cv_swap_.notify_one();
    }
}

bool diffraflow::Sender::swap_() {
    unique_lock<mutex> lk(mtx_);
    if (buffer_A_limit_ < size_threshold_) {
        cv_swap_.wait_for(lk, std::chrono::microseconds(time_threshold_));
    }
    if (buffer_A_limit_ > 0) {
        // do the swap
        char*  tmp_buff = buffer_A_;
        size_t tmp_size = buffer_A_limit_;
        buffer_A_ = buffer_B_;
        buffer_A_limit_ = buffer_B_limit_;
        buffer_B_ = tmp_buff;
        buffer_B_limit_ = tmp_size;
        return true;
    } else {
        return false;
    }
}

void diffraflow::Sender::send_() {
    // try to connect if lose connection
    if (client_sock_fd_ < 0) {
        if (connect_to_combiner()) {
            // BOOST_LOG_TRIVIAL(info) << "reconnected to combiner.";
        } else {
            // BOOST_LOG_TRIVIAL(warning) << "failed to reconnect to combiner, discard data in buffer.";
            buffer_B_limit_ = 0;
            return;
        }
    }
    for (size_t pos = 0; pos < buffer_B_limit_;) {
        int count = write(client_sock_fd_, buffer_B_ + pos, buffer_B_limit_ - pos);
        if (count == 0) {
            // BOOST_LOG_TRIVIAL(warning) << "connection is closed from the other side.";
            close_connection();
            buffer_B_limit_ = 0;
            return;
        } else if (count < 0) {
            BOOST_LOG_TRIVIAL(warning) << "error found when sending data: " << strerror(errno);
            buffer_B_limit_ = 0;
            return;
        } else {
            pos += count;
        }
    }
}

void diffraflow::Sender::run_() {
    while (run_flag_) {
        if (swap_()) send_();
    }
}

void diffraflow::Sender::start() {

}

void diffraflow::Sender::stop() {

}
